OAK-10048: DocumentStoreIndexerBase#buildFlatFileStore outputs the wrong path when FlatFileSplitter used

amit-jain opened the pull request.

Commit:
OAK-10048: DocumentStoreIndexerBase#buildFlatFileStore outputs the wrong path when FlatFileSplitter used

- System property (oak.indexer.sortedFilePath) defines the folder of the flat file store rather than the current specific file path
- Enable buildFlatFiles to use the folder in the system property defined as the place to look for files
- In case the folder is split then it uses that to init the split flat file store

Commit:
Merge branch 'trunk' into OAK-10048

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:
```diff
@@ -222,26 +235,53 @@ public List<FlatFileStore> buildList(IndexHelper indexHelper, IndexerSupport ind
+        // Initialize the flat file store again
+        
+        createStoreDir();
+        org.apache.jackrabbit.oak.index.indexer.document.flatfile.SortStrategy strategy = createSortStrategy(flatFileStoreDir);
```
Couldn't we import the class?

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:
```diff
@@ -222,26 +235,53 @@ public List<FlatFileStore> buildList(IndexHelper indexHelper, IndexerSupport ind
+        File sortedDir = new File(sortedFilePath);
+        if (sortedDir.exists() && sortedDir.canRead() && sortedDir.isDirectory()) {
+            log.info("Reading from provided sorted files directory [{}] (via system property '{}')",
+                sortedDir.getAbsolutePath(), OAK_INDEXER_SORTED_FILE_PATH);
```
The log message might be a bit misleading... here you write "(via system property x)" but in fact the caller used that system property (createdSortedStoreFiles()), not this class... What about logging in the caller about the system property value, and here just "Reading from provided sorted files directory [{}]"?

This will log twice, but I think that's fine: it is more explicit on what it is doing.

amit-jain replied:
The caller (createdSortedStoreFiles()) as well as the definition of the system property is in the same class.

amit-jain replied:
I moved the usage of OAK_INDEXER_SORTED_FILE_PATH to the caller.

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:
```diff
@@ -222,26 +235,53 @@ public List<FlatFileStore> buildList(IndexHelper indexHelper, IndexerSupport ind
+     */
+    private List<File> createdSortedStoreFiles() throws IOException, CompositeException {
+        // Check system property defined path
         String sortedFilePath = System.getProperty(OAK_INDEXER_SORTED_FILE_PATH);
```
(I have to admit I don't understand why we use a system property, and not a parameter... but this is not the fault of this patch, so let's keep that for now.)

reschke replied:
FWFW, if it needs to be a system property, it should use https://jackrabbit.apache.org/oak/docs/apidocs/org/apache/jackrabbit/oak/commons/properties/SystemPropertySupplier.html

thomasmueller replied:
SystemPropertySupplier is nice because it logs if the value is set (to a non-default value). However in this case I think not using it is fine.

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:
```diff
@@ -31,13 +32,16 @@
 import org.slf4j.LoggerFactory;
 
 import java.io.File;
+import java.io.FilenameFilter;
```
not used

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java:
```diff
@@ -73,7 +74,9 @@ public class FlatFileSplitter {
     private boolean useLZ4 = Boolean.parseBoolean(System.getProperty(OAK_INDEXER_USE_LZ4, "false"));
-
+    
+    static Predicate IS_SPLIT = (Predicate<File>) path -> path.getParent().endsWith(SPLIT_DIR_NAME);
```
I would change this with the following
```suggestion
    static Predicate<File> IS_SPLIT = path -> path.getParent().endsWith(SPLIT_DIR_NAME);
```

fabriziofortino commented on oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilderTest.java:
```diff
@@ -60,60 +87,162 @@ public void defaultSortStrategy() throws Exception {
+            assertTrue(storeList.size() > 1);
+        } else {
+            assertEquals(dir, storeList.get(0).getFlatFileStorePath());
+            assertTrue(storeList.size() == 1);
```
it can be simplified
```suggestion
            assertEquals(1, storeList.size());
```

fabriziofortino reviewed (commented):
a few minors

Commit:
OAK-10048: DocumentStoreIndexerBase#buildFlatFileStore outputs the wrong path when FlatFileSplitter used

Review changes:
- Removing unneeded import
- Cleaning up logging
- cleaned up predicate init
- simplified assertion in tests
